### PR TITLE
Allow to dry run the register component operation

### PR DESCRIPTION
**Pull request by Fox32**
This adds the ability to dry run adding a new location ot the catalog API. The location is now added in a transaction and afterwards rolled back. This allows users to dry run this operation to see if there entity has issues.
In the future this can be done by automated tools in the CI/CD pipeline, like GitHub actions.

Also adds a validate button to the register-component page This allows the user to validate his location before adding it.

image

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/spotify/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)

### CODE REVIEW

File: plugins/catalog-backend/src/ingestion/HigherOrderOperations.ts
Comment (Fox32) [inline plugins/catalog-backend/src/ingestion/HigherOrderOperations.ts:91]: The higher order operation uses multiple database calls (matches it's name? 😆 ). That means we have to move the transaction up to here.

File: plugins/catalog-backend/src/service/router.ts
Comment (Fox32) [inline plugins/catalog-backend/src/service/router.ts:89]: My first draft also added dryRun to `POST /entities` but I don't need it right now and I think this way it is sufficient.

**Comment by freben** [comment]
sorry to say that a refactor collided with you - there are some conflicts

**Comment by Fox32** [comment]
No problem, I will take a look tomorrow :)

File: plugins/catalog-backend/src/catalog/DatabaseEntitiesCatalog.ts
Comment (freben) [inline plugins/catalog-backend/src/catalog/DatabaseEntitiesCatalog.ts:86]: there's `private`! :)

File: plugins/catalog-backend/src/ingestion/HigherOrderOperations.ts
Comment (freben) [reply plugins/catalog-backend/src/ingestion/HigherOrderOperations.ts:91]: Myeah ... it's a bit unfortunate but I don't actually have a better solution to offer right now. May want to refactor away the entire HigherOrderOperations class at some point but not right now.

File: plugins/catalog-backend/src/service/router.ts
Comment (freben) [inline plugins/catalog-backend/src/service/router.ts:91]: could perhaps use the `yn` lib here to support more truthy syntaxes, but not really necessary :P

File: plugins/register-component/src/components/RegisterComponentForm/RegisterComponentForm.test.tsx
Comment (freben) [inline plugins/register-component/src/components/RegisterComponentForm/RegisterComponentForm.test.tsx:23]: -`a`? :P

**Review by freben — APPROVED** [review]
Just minor comments. Approving pending a rebase

File: plugins/catalog-backend/src/catalog/DatabaseEntitiesCatalog.ts
Comment (Fox32) [reply plugins/catalog-backend/src/catalog/DatabaseEntitiesCatalog.ts:86]: To much Dart code 😆

File: plugins/catalog-backend/src/catalog/types.ts
Comment (Rugvip) [inline plugins/catalog-backend/src/catalog/types.ts:26]: https://github.com/spotify/backstage/pull/2617 removes this and `addEntities`, just keeping the batch version.

This layer really shouldn't care about transactions though, better just add a more explicit `dryRunEntityUpdate` kinda thing instead imo

File: plugins/catalog-backend/src/database/types.ts
Comment (Rugvip) [inline plugins/catalog-backend/src/database/types.ts:95]: 👌

File: plugins/catalog-backend/src/catalog/types.ts
Comment (Fox32) [reply plugins/catalog-backend/src/catalog/types.ts:26]: Yeah, as the relationship code is already open for a long time, finish that one first. I will rebase my code afterwards! :)

**Comment by benjdlambert** [comment]
Hey @Fox32!

So as you might have seen we have a lot of work internally happening on relations over in this PR #2617. That should hopefully be merged in the next day or so, but that could possibly cause some conflicts. I'd recommend holding off on this until we've merged our stuff and then you should be able to fix up this :)

**Comment by benjdlambert** [comment]
@Fox32 our upstream has been merged, so there's a few conflicts that's popped up in this PR.

File: plugins/register-component/src/components/RegisterComponentForm/RegisterComponentForm.test.tsx
Comment (Fox32) [reply plugins/register-component/src/components/RegisterComponentForm/RegisterComponentForm.test.tsx:23]: 😉

File: plugins/catalog-backend/src/service/router.ts
Comment (Fox32) [reply plugins/catalog-backend/src/service/router.ts:91]: First I thought: another dependency? but it's already there so: yes yeah!

File: plugins/catalog-backend/package.json
Comment (Fox32) [inline plugins/catalog-backend/package.json:35]: There was an noisy warning during testing, but its fixed with the update.

There is also a warning from ts-jest that could be resolved with an update, but that update fails somewhere else

File: plugins/catalog-backend/src/catalog/types.ts
Comment (Fox32) [reply plugins/catalog-backend/src/catalog/types.ts]: I would with that HigherOrderOperations would have no access to the database class, but that is not possible as the transaction has to contain changes to both the EntitiesCatalog and LocationsCatalog

File: plugins/catalog-backend/package.json
Comment (Fox32) [reply plugins/catalog-backend/package.json:35]: I could fix the ts-jest problem too, the remaining issues are unrelated to my changes?

**Comment by Fox32** [comment]
Nevermind, I was the one that updates knex...

File: plugins/catalog-backend/package.json
Comment (Fox32) [reply plugins/catalog-backend/package.json:35]: The ts-jest problem ist fixed to 😉

Comment (Fox32) [reply plugins/catalog-backend/package.json]: Had to revert both changes, looks like they have issues in tests somewhere else 🤔

**Comment by Fox32** [comment]
@benjdlambert  Build is green. A lot of things have changed, so I would prefer a new review. I would also prefer not to rebase it another time 😆

File: plugins/catalog-backend/src/ingestion/HigherOrderOperations.ts
Comment (Rugvip) [inline plugins/catalog-backend/src/ingestion/HigherOrderOperations.ts:98]: Still not too happy about exposing transactions at this layer. I think the main issue here is that we need to do the batch write and then read without modifying the DB? Wouldn't mind just adding any kind of `dryRunBatchAdd`-kinda method to the `EntitiesCatalog` for now. And then during a dry run we can easily just skip adding the location too right, so no need for transactions here.

We could also just make `batchAddOrUpdateEntities` a bit more powerful, giving it both a `dryRun` and a `output` options so that we don't even need the separate `entities` step below. That'd give us a place to solve both large entity volume issues and a method that can handle all of these use-cases and more.

Looking forward I think the `HigherOrderOperations` (or whatever we rename it to 😅 ) could also use its own `dryRun` method, something like

```ts
processEntities(opts: { entites: Entity[] }): EntityProcessingOutput
```

where `EntityProcessingOutput` (or w/e name) could eventually be a elaborate description on what effect processing the provided entities would have on the catalog, but right now it would just mirror the current `outputEntities` we're returning.

With that kind of method I think we'd be able to cover PR-feedback use-cases, as well as this dry-run use-case.

Dry-running a location `target` would essentially just be

```ts
const output = await catalog.processEntities({
  entities: [{
    apiVersion: 'backstage.io/v1alpha1',
    kind: 'Location',
    spec: {
      type: 'url'
      target,
    },
  }],
})
```

Thoughts on that @freben @benjdlambert, and think it's a sane direction @Fox32? 

Just to reiterate I'm thinking the best way forward here for now is to add both a `dryRun` option and an option to make `batchAddOrUpdateEntities` read back the entities it wrote.

Comment (Fox32) [reply plugins/catalog-backend/src/ingestion/HigherOrderOperations.ts:98]: I'm not sure what kind of errors see would miss if we don't add the location, but I agree that skipping that part allows to remove the database from here. I will give it a try.

File: plugins/catalog-backend/package.json
Comment (Fox32) [reply plugins/catalog-backend/package.json:35]: I will keep the knex update, but the ts-jest one still has problems.

**Review by OrkoHunter — APPROVED** [review]

**Review by Rugvip — APPROVED** [review]

**Pull request merged**
